Make TorchScript Preserve Fully Qualified Class Name for Python Exceptions

shunting314 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #70339

When a python program is translated to TorchScript, the python exception type is dropped. This makes users's life hard when they need to categorize errors based more than only exception message.

Here we make the change so when we raise a python exception, we record the fully qualified class name for the exception. Later on when the TorchScript is interpreted, a special exception CustomJITException is thrown. User can get the python class name from CustomJITException::getPythonClassName .

Note that, this diff does not customize the mapping from C++ exception to Python exception. It's left to the users to do whatever mapping they want.

Code under scripts/shunting are just my own experimental code. I can split them out if requested.

Differential Revision: [D33282878](https://our.internmc.facebook.com/intern/diff/D33282878/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D33282878/)!

shunting314 committed:
Make TorchScript Preserve Fully Qualified Class Name for Python Exceptions

When a python program is translated to TorchScript, the python exception type is dropped. This makes users's life hard when they need to categorize errors based more than only exception message.

Here we make the change so when we raise a python exception, we record the fully qualified class name for the exception. Later on when the TorchScript is interpreted, a special exception CustomJITException is thrown. User can get the python class name from CustomJITException::getPythonClassName .

Note that, this diff does not customize the mapping from C++ exception to Python exception. It's left to the users to do whatever mapping they want.

Code under scripts/shunting are just my own experimental code. I can split them out if requested.

Differential Revision: [D33282878](https://our.internmc.facebook.com/intern/diff/D33282878/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D33282878/)!

[ghstack-poisoned]

shunting314 committed:
Update on "Make TorchScript Preserve Fully Qualified Class Name for Python Exceptions"


When a python program is translated to TorchScript, the python exception type is dropped. This makes users's life hard when they need to categorize errors based more than only exception message.

Here we make the change so when we raise a python exception, we record the fully qualified class name for the exception. Later on when the TorchScript is interpreted, a special exception CustomJITException is thrown. User can get the python class name from CustomJITException::getPythonClassName .

Note that, this diff does not customize the mapping from C++ exception to Python exception. It's left to the users to do whatever mapping they want.

Code under scripts/shunting are just my own experimental code. I can split them out if requested.

Differential Revision: [D33282878](https://our.internmc.facebook.com/intern/diff/D33282878/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D33282878/)!

[ghstack-poisoned]

gmagogsfm commented on torch/csrc/jit/runtime/jit_exception.h:
```diff
@@ -11,5 +12,18 @@ struct TORCH_API JITException : public std::runtime_error {
   explicit JITException(const std::string& msg);
 };
 
+struct TORCH_API CustomJITException : public JITException {
```
Can we enhance `JITException` to have an `optional<str>` field instead of subclassing? Edge use cases are sensitive to binary size increase and runtime overhead, and subclassing would add to both.

gmagogsfm commented on test/jit/test_exception.py:
```diff
@@ -0,0 +1,166 @@
+        def fn():
+            raise MyValueError("test custom exception")
+
+        with self.assertRaisesRegex(torch.jit.Error, "__torch__.jit.test_exception.MyValueError: test custom exception"):
```
Can we test a case where exception is defined outside of `main`? TorchScript handles classes in `main` in a special case by adding a `__torch__` prefix. This is not common for users. It is more useful to guarantee that we can maintain the correct original fully qualified name for exceptions defined in non-main modules. could we add a test for that?

shunting314 replied:
The exception is defined in jit.test_exception module actually. Do you actually want me create an exception in a different module to the one raising the exception?

gmagogsfm replied:
Yeah, let's try defining it in a different module and raise it to see if qualified name is easily recognizable.

shunting314 replied:
Just tried. The __torch__ prefix is added when the exception module is added elsewhere as well. Looks like here is the code adding the prefix: https://github.com/pytorch/pytorch/blob/d100d98db8424d6d88c9d72d156c49ba94e9d18e/torch/_jit_internal.py#L1029-L1035 .

Will send a update with the extra unit tests after I resolve other comments.

gmagogsfm replied:
I see. This should be fine, as long as we are not losing original module_name. For `main`, the original module name is lost.

gmagogsfm commented on torch/csrc/jit/python/python_sugared_value.h:
```diff
@@ -340,6 +343,9 @@ struct VISIBILITY_HIDDEN PythonExceptionValue : public ExceptionValue {
       size_t n_binders) override;
+
+ private:
+  std::string qualified_name_;
```
s/qualified_name_/exception_class_qualified_name/

gmagogsfm commented:
Looks good, just some minor comments.

shunting314 committed:
Update on "Make TorchScript Preserve Fully Qualified Class Name for Python Exceptions"


When a python program is translated to TorchScript, the python exception type is dropped. This makes users's life hard when they need to categorize errors based more than only exception message.

Here we make the change so when we raise a python exception, we record the fully qualified class name for the exception. Later on when the TorchScript is interpreted, a special exception CustomJITException is thrown. User can get the python class name from CustomJITException::getPythonClassName .

Note that, this diff does not customize the mapping from C++ exception to Python exception. It's left to the users to do whatever mapping they want.

Code under scripts/shunting are just my own experimental code. I can split them out if requested.

Differential Revision: [D33282878](https://our.internmc.facebook.com/intern/diff/D33282878/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D33282878/)!

[ghstack-poisoned]

shunting314 committed:
Update on "Make TorchScript Preserve Fully Qualified Class Name for Python Exceptions"


When a python program is translated to TorchScript, the python exception type is dropped. This makes users's life hard when they need to categorize errors based more than only exception message.

Here we make the change so when we raise a python exception, we record the fully qualified class name for the exception. Later on when the TorchScript is interpreted, a special exception CustomJITException is thrown. User can get the python class name from CustomJITException::getPythonClassName .

Note that, this diff does not customize the mapping from C++ exception to Python exception. It's left to the users to do whatever mapping they want.

Code under scripts/shunting are just my own experimental code. I can split them out if requested.

Differential Revision: [D33282878](https://our.internmc.facebook.com/intern/diff/D33282878/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D33282878/)!

[ghstack-poisoned]

shunting314 committed:
Update on "Make TorchScript Preserve Fully Qualified Class Name for Python Exceptions"


When a python program is translated to TorchScript, the python exception type is dropped. This makes users's life hard when they need to categorize errors based more than only exception message.

Here we make the change so when we raise a python exception, we record the fully qualified class name for the exception. Later on when the TorchScript is interpreted, a special exception CustomJITException is thrown. User can get the python class name from CustomJITException::getPythonClassName .

Note that, this diff does not customize the mapping from C++ exception to Python exception. It's left to the users to do whatever mapping they want.

Code under scripts/shunting are just my own experimental code. I can split them out if requested.

Differential Revision: [D33282878](https://our.internmc.facebook.com/intern/diff/D33282878/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D33282878/)!

[ghstack-poisoned]

shunting314 committed:
Update on "Make TorchScript Preserve Fully Qualified Class Name for Python Exceptions"


When a python program is translated to TorchScript, the python exception type is dropped. This makes users's life hard when they need to categorize errors based more than only exception message.

Here we make the change so when we raise a python exception, we record the fully qualified class name for the exception. Later on when the TorchScript is interpreted, a special exception CustomJITException is thrown. User can get the python class name from CustomJITException::getPythonClassName .

Note that, this diff does not customize the mapping from C++ exception to Python exception. It's left to the users to do whatever mapping they want.

Code under scripts/shunting are just my own experimental code. I can split them out if requested.

Differential Revision: [D33282878](https://our.internmc.facebook.com/intern/diff/D33282878/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D33282878/)!

[ghstack-poisoned]

gmagogsfm commented on torch/csrc/jit/python/python_sugared_value.h:
```diff
@@ -328,7 +328,10 @@ struct VISIBILITY_HIDDEN PythonClassValue : public ClassValue {
       : ExceptionValue(
-            py::str(py::getattr(exception_class, "__name__", py::str("")))) {}
+            py::str(py::getattr(exception_class, "__name__", py::str("")))) {
+    exception_class_qualified_name_ = py::str(py::module::import("torch._jit_internal")
+                                  .attr("_qualified_name")(exception_class));
```
super nit:
This can be rewritten in to member initialization list like `ExceptionValue`.

gmagogsfm replied:
Actually maybe it is better if we use the un-mangled qualified name here. Since this is just going to be a string in the IR/output. There is no chance it is going to be confused with classes from other compilation units. What do you think?

shunting314 replied:
Sure, I'll add an optional argument to torch._jit_internal._qualified_name to control if mangling

shunting314 committed:
Update on "Make TorchScript Preserve Fully Qualified Class Name for Python Exceptions"


When a python program is translated to TorchScript, the python exception type is dropped. This makes users's life hard when they need to categorize errors based more than only exception message.

Here we make the change so when we raise a python exception, we record the fully qualified class name for the exception. Later on when the TorchScript is interpreted, a special exception CustomJITException is thrown. User can get the python class name from CustomJITException::getPythonClassName .

Note that, this diff does not customize the mapping from C++ exception to Python exception. It's left to the users to do whatever mapping they want.

Code under scripts/shunting are just my own experimental code. I can split them out if requested.

Differential Revision: [D33282878](https://our.internmc.facebook.com/intern/diff/D33282878/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D33282878/)!

[ghstack-poisoned]

gmagogsfm commented on torch/csrc/jit/python/python_sugared_value.h:
```diff
@@ -328,7 +328,10 @@ struct VISIBILITY_HIDDEN PythonClassValue : public ClassValue {
+            py::str(py::getattr(exception_class, "__name__", py::str("")))),
+        exception_class_qualified_name_(
+            py::str(py::module::import("torch._jit_internal")
+                        .attr("_qualified_name")(exception_class, false))) {}
```
```suggestion
                        .attr("_qualified_name")(exception_class, /*mangle_name=*/false))) {}
```

shunting314 committed:
Update on "Make TorchScript Preserve Fully Qualified Class Name for Python Exceptions"


When a python program is translated to TorchScript, the python exception type is dropped. This makes users's life hard when they need to categorize errors based more than only exception message.

Here we make the change so when we raise a python exception, we record the fully qualified class name for the exception. Later on when the TorchScript is interpreted, a special exception CustomJITException is thrown. User can get the python class name from CustomJITException::getPythonClassName .

Note that, this diff does not customize the mapping from C++ exception to Python exception. It's left to the users to do whatever mapping they want.

Code under scripts/shunting are just my own experimental code. I can split them out if requested.

Differential Revision: [D33282878](https://our.internmc.facebook.com/intern/diff/D33282878/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D33282878/)!

[ghstack-poisoned]

seemethere commented:
This was reverted in https://github.com/pytorch/pytorch/commit/bf610f08b07a1135eaffd1ee53a1b4aa901601c1
